File: typescript/src/refactoring/golden-master/golden-master.test.ts
import {describe, expect, it} from 'vitest'
import {generateOrder, ReceiptPrinter} from './golden-master'

describe('Receipt Printer', () => {
  let counter = 0
  describe('Given a customer', () => {
    const customer = 'Ana'
    describe('Given a number of items', () => {
      const item = 1
      describe('Given quantity', () => {
        const quantity = 1
        it('should print a receipt', () => {
          counter = counter + 1
          const order = generateOrder('ORD-' + counter.toString(), customer, item, quantity)
          expect(new ReceiptPrinter().print(order)).matchSnapshot()
        })
      })
    })
  })
})

File: java/src/test/java/com/refactoring/refactoring/GoldenMasterTest.java
package com.refactoring.refactoring;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.Date;

@ExtendWith(SnapshotExtension.class)
class GoldenMasterTest {

    private static int counter = 0;
    private Expect expect;

    @Test
    void shouldPrintAReceipt() {
        // Given a customer
        String customer = "Ana";
        // Given a number of items
        int item = 1;
        // Given quantity
        int quantity = 1;

        counter++;
        Order order = OrderGenerator.generateOrder("ORD-" + counter, customer, item, quantity);
        String receipt = new ReceiptPrinterWithoutDiscountForTest().print(order);

        expect.toMatchSnapshot(receipt);
    }

    static class ReceiptPrinterWithoutDiscountForTest extends ReceiptPrinter {
        @Override
        protected Date getCurrentDate() {
            return new Date(122, 1, 1); // Year 2022, Month February (0-indexed), Day 1
        }

        @Override
        protected double discount() {
            return 0.0;
        }
    }

    static class ReceiptPrinterWithDiscountForTest extends ReceiptPrinter {
        @Override
        protected Date getCurrentDate() {
            return new Date(122, 1, 1); // Year 2022, Month February (0-indexed), Day 1
        }

        @Override
        protected double discount() {
            return 0.05;
        }
    }
}
